### ISSUE TITLE
[HOLD for payment 2024-02-15] [HOLD for payment 2024-02-14] [$1000] HIGH: Show "Room description" everywhere (2/2, external)

### ISSUE THREAD

**Issue by quinthar** [Expensify/App#32615]
*This is held on #32953 because we want to refactor the confusing naming of "Room welcome message" to "Room description" before changing where that description shows up.*

**Strategy:**
Chat is the highest-frequency action that happens inside of a company -- whoever owns a company's chat, has the strategic high ground to cross-sell everything else, as that's the stickiest use case.  Part of chat involves getting a company to capture all its conversations on the platform, split up into a bunch of room.  Each room is devoted to a different topic, and the "room welcome message" is how the admin explains what that topic is.  It is currently delivered via a whisper when someone joins the room.

**Problem:**
When the admin configures the "room welcome message", they currently don't see it anywhere in product (because it's only delivered via whisper to new people when they are added, and by definition they are already in it because they created it.  So, this just feels broken.  Additionally, there's no way an admin would guess that it's going to be secretly delivered to new users when they join, which is confusing.

![Image](https://github.com/Expensify/App/assets/575186/ab4846b0-0ce6-43fb-a221-c152dbed003f)

**Solution:**
Update room welcome message to be room description and put the description at the top of the chat, in addition to delivering it via a whisper to those who join a room that already has enough content that the top has scrolled off the screen.  Specifically, a chat has:

* Header:
    * `<#room name> in <workspace name>` - pressing opens the chat details
    * `<workspace description; truncated>` - pressing opens the description editor in the RHP
        * Note: If there is no custom description; remove this 2nd line and vertically center the first
        * Note: Only workspace members who have write access to the room should be able to edit the description
* Chat masthead:
    * `Welcome to <#roomname>`
    * `<room description, full>` - pressing opens the description editor in the RHP
        * Note: If there is no custom description, use the current default masthead language
* Chat details:
    * Avatar
    * `<#room name>` - Any member can press to rename the report
    * `<workspace name>`
    * "Description"
    * `<room description>` - Shows the first 5 lines truncated, with a `more` control to expand fully;
        * Note: Only workspace members who have write access to the room should be able to edit the description
* API
    * Remove all `UpdateWelcomeMessage` API calls
    * Replace them with `updateRoomDescription` API calls

**Mockups:**
![Image](https://github.com/Expensify/App/assets/575186/a06296e4-b835-4975-8ea3-7d99dfca41f4)
![Image](https://github.com/Expensify/App/assets/575186/0118ed20-b62c-4672-b848-c78440d8c658)

More discussion [in Slack here](https://expensify.slack.com/archives/CC7NECV4L/p1702001845470919).
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ce65d2b9e9db71e4</li>
        <li>Upwork Job ID: 1744191782833340416</li>
        <li>Last Price Increase: 2024-01-15</li>
<li>Automatic offers: </li>
<ul>
<li>esh-g | Contributor | 28088918</li>
<li>ishpaul777 | Contributor | 28148450</li>
</ul></li>
    </ul>
</details>

**Comment by esh-g** [comment]
# Proposal

## Please re-state the problem we are trying to solve
Update the room description to be shown in the header and make it editable for admins

## What is the root cause of this problem
N/A (New feature)

## What changes should be made to fix this?
1. We need to change the welcome message in the main pane here:
https://github.com/Expensify/App/blob/d6f359fc010fe8120cf8a6761a4dac21c819e270/src/components/ReportWelcomeText.js#L92-L98
We should render this only when `roomDescription` doesn't exist
Then we can add a conditional to render for the `roomDescription` with  `<RenderHTML html={roomDescription} />` (assuming we want the description to have html capabilities, otherwise we can use `Text`)
We can change this line:
https://github.com/Expensify/App/blob/d6f359fc010fe8120cf8a6761a4dac21c819e270/src/components/ReportWelcomeText.js#L72
```js
const roomDescription = lodashGet(props.report, 'description');
...
{isChatRoom && (
<>
  {!_.isEmpty(roomDescription) ? <RenderHTML html={roomDescription} /> : (
      <>
          <Text>{roomWelcomeMessage.phrase1}</Text>
          {roomWelcomeMessage.showReportName && (
              <Text
                  style={[styles.textStrong]}
                  onPress={() => Navigation.navigate(ROUTES.REPORT_WITH_ID_DETAILS.getRoute(props.report.reportID))}
                  suppressHighlighting
              >
                  {ReportUtils.getReportName(props.report)}
              </Text>
          )}
          {roomWelcomeMessage.phrase2 !== undefined && <Text>{roomWelcomeMessage.phrase2}</Text>}
      </>
  )}
  
</>
)}
```
<img width="1030" alt="Screenshot 2023-12-27 at 10 12 28 PM" src="https://github.com/Expensify/App/assets/77237602/ba902b76-23bb-4e58-8d5b-d6b02a1772ee">


2. For changing the header to have the description as the subtitle, we first need to do the following:
- Create a function to get the description after sanitising all the html from the text (`ReportUtils.getChatRoomDescription()`)
- Change the conditional to display the current `subtitle` only when room description not present (`!_.isEmpty(subtitle) && _.isEmpty(roomDescription)`)
- Add a conditional render for room description after the subtitle. (`!_.isEmpty(roomDescription)`)
- Currently the title (`DisplayNames`) is nested in a flex column with the subtitle. We need to add another wrapper for `DisplayNames` block [here](https://github.com/Expensify/App/blob/d6f359fc010fe8120cf8a6761a4dac21c819e270/src/pages/home/HeaderView.js#L253) as a flex row with it's sibling being the text `in Workspace Co.`
So here are the changes:
```js
function getChatRoomDescription(report: Report) {
    if (!report.description) {
        return '';
    }
    // return after removing the html.
    const parser = new ExpensiMark();
    return parser.htmlToText(report.description);
}
```
```js
const roomDescription = ReportUtils.getChatRoomDescription(props.report);
const policyName = ReportUtils.getPolicyName(props.report);
...
<View style={[styles.flex1, styles.flexColumn]}>
    <View style={[styles.flex1, styles.flexRow, styles.gap1, styles.alignItemsBaseline]}> // add this row wrapper
        <DisplayNames
            fullTitle={title}
            displayNamesWithTooltips={displayNamesWithTooltips}
            tooltipEnabled
            numberOfLines={1}
            textStyles={[styles.headerText, styles.pre]}
            shouldUseFullTitle={isChatRoom || isPolicyExpenseChat || isChatThread || isTaskReport}
        /> 
        {isChatRoom && !_.isEmpty(policyName) && !_.isEmpty(roomDescription) && ( // code for the policyName display
            <>
                <Text
                    style={[styles.sidebarLinkText, styles.textLabelSupporting]}
                    numberOfLines={1}
                >
                    in {/* A translation can obviously be added. */}
                </Text>
                <Text 
                    style={[styles.sidebarLinkText, styles.textLabelSupporting, styles.textStrong]}
                    numberOfLines={1}
                >
                    {policyName}
                </Text>
            </>
        )}
    </View>
    ...
 ```
 ```js
{!_.isEmpty(roomDescription) && (
    <Text
        style={[styles.sidebarLinkText, styles.optionAlternateText, styles.textLabelSupporting]}
        numberOfLines={1}
    >
        {roomDescription}
    </Text>
)}
```
**Result**
<img width="707" alt="Screenshot 2023-12-28 at 8 19 34 AM" src="https://github.com/Expensify/App/assets/77237602/596e10ac-f64a-46e6-9d60-fa1bab315984">

3. We also need to add the room description item in `ReportDetails` page which can be done like the following (we also need to rename the `REPORT_WELCOME_MESSAGE` page to `REPORT_DESCRIPTION`):
```js
{isChatRoom && (
    <MenuItemWithTopDescription
        shouldShowRightIcon={isPolicyAdmin}
        title={props.report.welcomeMessage}
        shouldRenderAsHTML
        description={translate('common.description')}
        onPress={() => Navigation.navigate(ROUTES.REPORT_WELCOME_MESSAGE.getRoute(props.report.reportID))}
    />
)}
```
<img width="363" alt="Screenshot 2023-12-28 at 8 32 56 AM" src="https://github.com/Expensify/App/assets/77237602/98efcf68-edff-483c-b06f-1daa106a85de">

**Final Result**

https://github.com/Expensify/App/assets/77237602/d67420bb-78cc-4041-9ea4-36796f971291

(The thing pending is to rename all the `welcomeMessage` page and routes and translations to description)

**Comment by puneetlath** [comment]
@rushatgabhane looks like we already have one proposal. Please give it a review when you have a chance. And let me know if anything about the requirements is unclear.

**Comment by rushatgabhane** [comment]
@puneetlath i think this is pure implementation. We can assign @esh-g because they have demonstrated a working solution. 
We'll refine the code in the PR

🎀 👀 🎀

**Comment by puneetlath** [comment]
Sounds good. @esh-g when do you think you can have a PR ready for review?

**Comment by puneetlath** [comment]
BTW the `UpdateRoomDescription` API command isn't quite live on staging yet, but I'll ping here once it is. Should be sometime today.

**Comment by puneetlath** [comment]
Ok @esh-g the API command is on staging now, so you should be able to test your PR with it.

**Comment by esh-g** [comment]
@puneetlath Seems like the backend removes the html tags from the description when I set it in the request... I think it should be allowed. Could you please check that?

**Comment by esh-g** [comment]
@puneetlath We also need to modify the `AddWorkspaceRoom` API endpoint to accept a description instead of `welcomeMessage`

**Comment by esh-g** [comment]
The PR is here: https://github.com/Expensify/App/pull/34150

**Comment by puneetlath** [comment]
>@puneetlath Seems like the backend removes the html tags from the description when I set it in the request... I think it should be allowed. Could you please check that?

Hm, why should it support html tags? In order to support markdown? 

>@puneetlath We also need to modify the AddWorkspaceRoom API endpoint to accept a description instead of welcomeMessage

Good catch! I'll get that added.

**Comment by esh-g** [comment]
> Hm, why should it support html tags? In order to support markdown?
Yes @puneetlath I think it all markdown enabled text is stored as HTML in the backend, and I think we want to enable markdown for the description...

**Comment by esh-g** [comment]
Hi, I just wanted to know if it would be possible to re-consider the bounty of the issue since this involved a lot more changes than normal issues (but fine either way)...

**Comment by puneetlath** [comment]
What do you think is fair @esh-g, given the scope?

**Comment by quinthar** [comment]
What is the ETA for completion?

**Comment by esh-g** [comment]
@quinthar I'm still waiting for the review by @rushatgabhane on the [PR](https://github.com/Expensify/App/pull/34150)

**Comment by esh-g** [comment]
> What do you think is fair @esh-g, given the scope?

@puneetlath I think the previous bounty amount ($1000) would be more than fair, but I'm fine with any decision you make!

**Comment by puneetlath** [comment]
Looking at the PR, I think that is also fair. I'll update.

**Comment by esh-g** [comment]
ETA Update:
Waiting on some additional backend changes to be made (should take 1-2 days)

**Comment by quinthar** [comment]
What exactly are the backend changes you are waiting on, and who is doing them -- is that @puneetlath?

**Comment by esh-g** [comment]
@quinthar The backend changes pending are:
- Remove `html` field from the description and store it as a `string` instead of `{html: string}`
- make sure the html tags passed from frontend are stored as they are sent (save html)

**Comment by puneetlath** [comment]
@esh-g that's on staging now.

**Comment by quinthar** [comment]
Great! Now that we save it as HTML, are we we using a markdown editor on the front end?  (In other words, when you edit it, do we convert the HTML to markdown, and when you save do we convert the markdown to HTML?)

**Comment by quinthar** [comment]
Regardless, what's the ETA for this getting done?

**Comment by quinthar** [comment]
Great!  Still on track to finish the PR today?

**Comment by puneetlath** [comment]
@rushatgabhane when will you be able to review the PR? Can you do it today?

**Comment by quinthar** [comment]
Ya, we've got a lot of value ready to go out the door, let's ship it!

**Comment by trjExpensify** [comment]
Just happened to stumble across this issue. Are we going to be doubling up the workspaceName in the header for workspace chats (member view)? 😕 

<img width="426" alt="image" src="https://github.com/Expensify/App/assets/16232057/209a3bbb-1d21-4043-af82-f9425054901b">

**Comment by esh-g** [comment]
@trjExpensify Here is the issue relating to workspace chats: https://github.com/Expensify/App/issues/32955

**Comment by JmillsExpensify** [comment]
Doesn't Tom's question also apply to this issue? I notice this mention in the solution for this issue.

> * Header:
>    * `<#room name> in <workspace name>` - pressing opens the chat details
>   * `<workspace description; truncated>` - pressing opens the description editor in the RHP
>        * Note: If there is no custom description; remove this 2nd line and vertically center the first
>        * Note: Only workspace members who have write access to the room should be able to edit the description

So for the case of a workspace chat, I think this solution is saying that it'd be <workspace name> in <workspace name> for any admin looking at the workspace chat. Are we sure we want that?

**Comment by quinthar** [comment]
Can you rephrase the question?  What does "I think this solution is saying
that it'd be in for any admin looking at the workspace chat" mean?

**Comment by JmillsExpensify** [comment]
Sure thing. My interpretation of this solution is that all admins will see a workspace chat that has the header. 

 - `<workspace name>` in `<workspace name>`
 
 So an example could be `Expensify Inc. in Expensify Inc.`. That feels pretty repetitive so I'm questioning that this is the ideal solution.

**Comment by esh-g** [comment]
@JmillsExpensify Thanks for clearing up what you meant! Here are the two cases for `policyExpenseChat`:
- The personal `policyExpenseChat` has the same name as workspace and yes, it may be repetitive to have `<workspace name>` in `<workspace name>` in this case.
- The `policyExpenseChat` for another member of the policy has the name of the person so it would be `person123` in `<workspace-name>`, so in this case, it would be required.

Therefore, I think we can remove the `name` in `name` format for personal `policyExpenseChat`, this can be done while implementation for the `policyDescription` issue. Because in this issue, we don't change the header for `policyExpenseChat` reports.

**Comment by quinthar** [comment]
Sounds great!  What's your new ETA for when this will be completed?

**Comment by esh-g** [comment]
@quinthar there is a PR open here: https://github.com/Expensify/App/pull/34150

**Comment by puneetlath** [comment]
@esh-g has been paid. 

@ishpaul777 can you suggest regression test steps here? Once done, I'll pay you out too. Thanks!

**Comment by puneetlath** [comment]
Great, thanks. All paid. 

Thanks y'all!
